### PR TITLE
autocomplete: Add and test MentionAutocompleteQuery

### PR DESCRIPTION
**Pull request by chrisbobbe**
This is atop my current revision of #84. It excludes much of the work we did in the office together, which I still have in a draft branch (and which I have an open question about: [czo](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/flutter.3A.20autocomplete.20computations/near/1559416))—but this seems like a self-contained piece of logic that we'll need, and that we'll want to get right.

Related: #49

### CODE REVIEW

File: lib/model/autocomplete.dart [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import '../api/model/model.dart';
+
+class MentionAutocompleteQuery {
+  MentionAutocompleteQuery(this.raw)
+    : _lowercaseWords = raw.toLowerCase().split(' ');
+
+  final String raw;
+
+
```
Comment (gnprice) [inline lib/model/autocomplete.dart:9]:
```suggestion

```

File: test/model/autocomplete_test.dart [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    doCheck('Name Words', eg.user(fullName: 'Full Name Four Words'), true);
+
+    doCheck('Fully Named', eg.user(fullName: 'Full Name'), false);
+    doCheck('Full Name', eg.user(fullName: 'Full'), false);
+    doCheck('Full Name', eg.user(fullName: 'Name'), false);
+    doCheck('ull ame', eg.user(fullName: 'Full Name'), false);
+    doCheck('ull Name', eg.user(fullName: 'Full Name'), false);
+    doCheck('Full ame', eg.user(fullName: 'Full Name'), false);
+    doCheck('Full Full', eg.user(fullName: 'Full Name'), false);
+    doCheck('Name Name', eg.user(fullName: 'Full Name'), false);
```
Comment (gnprice) [inline test/model/autocomplete_test.dart:32]: Also test that "full full" *does* match "Full Full Name" and "Full Name Full".

File: test/model/autocomplete_test.dart [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    doCheck('', eg.user(fullName: 'Full Name'), true);
+    doCheck('Full Name', eg.user(fullName: 'Full Name'), true);
+    doCheck('full name', eg.user(fullName: 'Full Name'), true);
+    doCheck('Full Name', eg.user(fullName: 'full name'), true);
+    doCheck('Full', eg.user(fullName: 'Full Name'), true);
+    doCheck('Name', eg.user(fullName: 'Full Name'), true);
+    doCheck('Full Name', eg.user(fullName: 'Fully Named'), true);
+    doCheck('Full Four', eg.user(fullName: 'Full Name Four Words'), true);
```
Comment (gnprice) [inline test/model/autocomplete_test.dart:22]: Also test that "F Full" and "Four F" don't match this name, while "Full F" and "F Four" do.

File: test/model/autocomplete_test.dart [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import '../example_data.dart' as eg;
+
+void main() {
+  test('MentionAutocompleteQuery.testUser', () {
+    doCheck(String rawQuery, User user, bool expected) {
+      final result = MentionAutocompleteQuery(rawQuery).testUser(user);
+      expected ? check(result).isTrue() : check(result).isFalse();
+    }
+
+    doCheck('', eg.user(fullName: 'Full Name'), true);
```
Comment (gnprice) [inline test/model/autocomplete_test.dart:15]: Also include tests with empty `fullName` ­— one where query is also empty, one where nonempty.

That isn't a super likely case, but it wouldn't be hard to have a bug where the search crashed on it.

**Review by gnprice — COMMENTED** [review]
Thanks! Small comments.

File: test/example_data.dart [hunk 5]
```diff
@@ -71,6 +71,30 @@ StreamMessage streamMessage(
...
 }
 
+User user({int? userId, String? fullName}) {
+  // The use of JSON here is convenient in order to delegate parts of the data
+  // to helper functions.  The main downside is that it loses static typing
+  // of the properties as we're constructing the data.  That's probably OK
+  // because (a) this is only for tests; (b) the types do get checked
+  // dynamically in the constructor, so any ill-typing won't propagate further.
+  return User.fromJson({
+    'user_id': userId ?? 123, // TODO generate example IDs
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply test/example_data.dart:81]: Ah yeah, I see I copied from `streamMessage` too hastily. 🙂

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed (with the commits from #84 refreshed with my current revision of that).

File: test/example_data.dart [hunk 6]
```diff
@@ -2,6 +2,27 @@ import 'package:zulip/api/model/initial_snapshot.dart';
...
+    isBot: false,
+    role: 400,
+    timezone: 'UTC',
+    avatarUrl: null,
+    avatarVersion: 0,
+    profileData: null,
+  );
+}
+
 final Uri realmUrl = Uri.parse('https://chat.example/');
```
Comment (gnprice) [inline test/example_data.dart:26]: nit: user data after realm-wide / server-wide data

**Review by gnprice — COMMENTED** [review]
One nit; otherwise lgtm once #84 is ready.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.
